GWT 2.7 RC1 : java.lang.ClassNotFoundException: com$gwtplatform$dispatch$rest$client$RestGinModule

When I use GWT 2.7RC1 and include:

```
<inherits name='com.gwtplatform.dispatch.rest.DispatchRest'/>
```
I did what was on this page (https://github.com/ArcBees/GWTP/wiki/Rest-Dispatch#add-rest-dispatch-to-your-gwtxml) in the important note.

I get the following error will running SuperDevMode and incremental compile:

```
GET /recompile/mobilePhoneGapIOS
   Job test.mobile.MobilePhoneGapIOS_1_18
      starting job: test.mobile.MobilePhoneGapIOS_1_18
      binding: mgwt.density=xhigh
      Compiling module test.mobile.MobilePhoneGapIOS
         Computing all possible rebind results for 'com.googlecode.gwtphonegap.client.log.shared.PhoneGapLogService'
            Rebinding com.googlecode.gwtphonegap.client.log.shared.PhoneGapLogService
               Invoking generator com.googlecode.gwtphonegap.rebind.PhoneGapServiceInterfaceProxyGenerator
                  Generating client proxy for remote service interface 'com.googlecode.gwtphonegap.client.log.shared.PhoneGapLogService'
                     Analyzing 'com.googlecode.gwtphonegap.client.log.shared.PhoneGapLogService' for serializable types
                        Analyzing methods:
                           public abstract java.lang.String logOnServer(java.lang.String clientId, java.util.List<java.util.logging.LogRecord> record)
                              Parameter: java.util.List<java.util.logging.LogRecord> record
                                 java.util.List<java.util.logging.LogRecord>
                                    Verifying instantiability
                                       com.google.common.collect.SingletonImmutableList<java.util.logging.LogRecord>
                                          Checking parameters of 'com.google.common.collect.SingletonImmutableList<java.util.logging.LogRecord>'
                                             Checking type argument 0 of type 'com.google.common.collect.SingletonImmutableList<E>' because it is directly exposed in this type or in one of its subtypes
                                                java.util.logging.LogRecord
                                                   Verifying instantiability
                                                      java.util.logging.LogRecord
                                                         Analyzing the fields of type 'java.util.logging.LogRecord' that qualify for serialization
                                                            private java.lang.Throwable thrown
                                                               java.lang.Throwable
                                                                  Verifying instantiability
                                                                     com.gwtplatform.dispatch.client.interceptor.InterceptorMismatchException
                                                                        Analyzing the fields of type 'com.gwtplatform.dispatch.client.interceptor.InterceptorMismatchException' that qualify for serialization
                                                                           [WARN] Field 'private final java.lang.Class<?> requestedActionType' will not be serialized because it is final
                                                                           [WARN] Field 'private final java.lang.Class<?> supportedActionType' will not be serialized because it is final
         Computing all possible rebind results for 'com.gwtplatform.mvp.client.DesktopGinjector'
            Rebinding com.gwtplatform.mvp.client.DesktopGinjector
               Invoking generator com.google.gwt.inject.rebind.GinjectorGenerator
                  [ERROR] Unable to load gin module type [com.gwtplatform.dispatch.rest.client.RestGinModule], maybe you haven't compiled your client java sources?
java.lang.ClassNotFoundException: com$gwtplatform$dispatch$rest$client$RestGinModule
	at com.google.gwt.inject.rebind.GinBridgeClassLoader.findClass(GinBridgeClassLoader.java:150)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
	at com.google.gwt.inject.rebind.GinBridgeClassLoader.loadClass(GinBridgeClassLoader.java:108)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:270)
	at com.google.gwt.inject.rebind.GinjectorGenerator.loadClass(GinjectorGenerator.java:223)
	at com.google.gwt.inject.rebind.GinjectorGenerator.getPropertyModuleClasses(GinjectorGenerator.java:137)
	at com.google.gwt.inject.rebind.GinjectorGenerator.getModuleClasses(GinjectorGenerator.java:116)
	at com.google.gwt.inject.rebind.GinjectorGenerator.generate(GinjectorGenerator.java:72)
	at com.google.gwt.core.ext.IncrementalGenerator.generateNonIncrementally(IncrementalGenerator.java:40)
	at com.google.gwt.dev.javac.StandardGeneratorContext.runGeneratorIncrementally(StandardGeneratorContext.java:760)
	at com.google.gwt.dev.cfg.RuleGenerateWith.realize(RuleGenerateWith.java:160)
	at com.google.gwt.dev.shell.StandardRebindOracle$Rebinder.rebind(StandardRebindOracle.java:79)
	at com.google.gwt.dev.shell.StandardRebindOracle.rebind(StandardRebindOracle.java:276)
	at com.google.gwt.dev.shell.StandardRebindOracle.rebind(StandardRebindOracle.java:265)
	at com.google.gwt.dev.DistillerRebindPermutationOracle.getAllPossibleRebindAnswers(DistillerRebindPermutationOracle.java:87)
	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.createStaticRebindExpression(UnifyAst.java:485)
	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.createRebindExpression(UnifyAst.java:443)
	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.handleMagicMethodCall(UnifyAst.java:576)
	at com.google.gwt.dev.jjs.impl.UnifyAst$UnifyVisitor.endVisit(UnifyAst.java:306)
	at com.google.gwt.dev.jjs.ast.JMethodCall.traverse(JMethodCall.java:248)
	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:381)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:293)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:285)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:128)
	at com.google.gwt.dev.jjs.ast.JCastOperation.traverse(JCastOperation.java:67)
	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:381)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:293)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:285)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:128)
	at com.google.gwt.dev.jjs.ast.JReturnStatement.traverse(JReturnStatement.java:40)
	at com.google.gwt.dev.jjs.ast.JModVisitor$ListContext.traverse(JModVisitor.java:95)
	at com.google.gwt.dev.jjs.ast.JModVisitor.acceptWithInsertRemove(JModVisitor.java:351)
	at com.google.gwt.dev.jjs.ast.JBlock.traverse(JBlock.java:92)
	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:381)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:293)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:149)
	at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:145)
	at com.google.gwt.dev.jjs.ast.JMethodBody.traverse(JMethodBody.java:83)
	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:381)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:293)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:285)
	at com.google.gwt.dev.jjs.ast.JMethod.visitChildren(JMethod.java:600)
	at com.google.gwt.dev.jjs.ast.JMethod.traverse(JMethod.java:569)
	at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:381)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:293)
	at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:285)
	at com.google.gwt.dev.jjs.impl.UnifyAst.mainLoop(UnifyAst.java:1505)
	at com.google.gwt.dev.jjs.impl.UnifyAst.exec(UnifyAst.java:870)
	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler$Precompiler.unifyJavaAst(JavaToJavaScriptCompiler.java:1303)
	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler$Precompiler.constructJavaAst(JavaToJavaScriptCompiler.java:1036)
	at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler$Precompiler.precompile(JavaToJavaScriptCompiler.java:953)
	at com.google.gwt.dev.jjs.MonolithicJavaToJavaScriptCompiler.precompile(MonolithicJavaToJavaScriptCompiler.java:303)
	at com.google.gwt.dev.jjs.JavaScriptCompiler.precompile(JavaScriptCompiler.java:38)
	at com.google.gwt.dev.Precompile.precompile(Precompile.java:286)
	at com.google.gwt.dev.Precompile.precompile(Precompile.java:229)
	at com.google.gwt.dev.Precompile.precompile(Precompile.java:145)
	at com.google.gwt.dev.Compiler.run(Compiler.java:206)
	at com.google.gwt.dev.codeserver.Recompiler.doCompile(Recompiler.java:323)
	at com.google.gwt.dev.codeserver.Recompiler.compile(Recompiler.java:164)
	at com.google.gwt.dev.codeserver.Recompiler.recompile(Recompiler.java:121)
	at com.google.gwt.dev.codeserver.Outbox.recompile(Outbox.java:128)
	at com.google.gwt.dev.codeserver.JobRunner.recompile(JobRunner.java:60)
	at com.google.gwt.dev.codeserver.JobRunner.access$100(JobRunner.java:31)
	at com.google.gwt.dev.codeserver.JobRunner$1.run(JobRunner.java:52)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
         [ERROR] Errors in 'gen/com/gwtplatform/mvp/client/DesktopGinjectorProvider.java'
            [ERROR] Line 8: Failed to resolve 'com.gwtplatform.mvp.client.DesktopGinjector' via deferred binding
         Unification traversed 7163 fields and methods and 1100 types. 518 are considered part of the current module and 518 had all of their fields and methods traversed.
      [ERROR] Compiler returned false
      [WARN] recompile failed
      [WARN] continuing to serve previous version

```

--- a/gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/annotations/Resource.java
+++ b/gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/annotations/Resource.java
@@ -1,0 +1,25 @@
+/*
+ * Copyright 2015 ArcBees Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.gwtplatform.dispatch.rest.client.annotations;
+
+/**
+ * Extend this interface from your own resource interfaces if you can't use the generated GIN module. If your project
+ * can use the generated GIN module (usually those using the generated entry point and ginjector from MvpWithEntryPoint)
+ * don't need to do this.
+ */
+public interface Resource {
+}

--- a/gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/core/CoreModule.java
+++ b/gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/core/CoreModule.java
@@ -40,5 +40,7 @@ public class CoreModule extends AbstractGinModule {
         bind(ResponseDeserializer.class).to(builder.getResponseDeserializer()).in(Singleton.class);
         bind(HttpParameterFactory.class).to(builder.getHttpParameterFactory()).in(Singleton.class);
         bind(RestDispatch.class).to(builder.getRestDispatch()).in(Singleton.class);
+
+        requestStaticInjection(StaticParametersFactory.class);
     }
 }

--- a/gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/core/StaticParametersFactory.java
+++ b/gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/client/core/StaticParametersFactory.java
@@ -1,0 +1,52 @@
+/*
+ * Copyright 2015 ArcBees Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.gwtplatform.dispatch.rest.client.core;
+
+import javax.inject.Inject;
+
+import com.gwtplatform.dispatch.rest.client.RestDispatch;
+import com.gwtplatform.dispatch.rest.client.annotations.DefaultDateFormat;
+import com.gwtplatform.dispatch.rest.client.core.parameters.HttpParameterFactory;
+
+/**
+ * Used internally by generated code to access various implementations bound by GIN. Injection could be used in
+ * generated code but because of changes introduced to the compiler in GWT 2.7, GIN generators may run before the
+ * generated GIN module is created and injection doesn't work in generated code.
+ * <p/>
+ * <strong>Don't use this class as it may change without notice.</strong>
+ */
+public class StaticParametersFactory {
+    @Inject
+    static RestDispatch restDispatch;
+    @Inject
+    static HttpParameterFactory httpParameterFactory;
+    @Inject
+    @DefaultDateFormat
+    static String defaultDateFormat;
+
+    public static RestDispatch getRestDispatch() {
+        return restDispatch;
+    }
+
+    public static HttpParameterFactory getHttpParameterFactory() {
+        return httpParameterFactory;
+    }
+
+    public static String getDefaultDateFormat() {
+        return defaultDateFormat;
+    }
+}

--- a/gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/rebind/FallThroughGenerator.java
+++ b/gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/rebind/FallThroughGenerator.java
@@ -1,0 +1,30 @@
+/*
+ * Copyright 2015 ArcBees Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.gwtplatform.dispatch.rest.rebind;
+
+import com.google.gwt.core.ext.Generator;
+import com.google.gwt.core.ext.GeneratorContext;
+import com.google.gwt.core.ext.TreeLogger;
+import com.google.gwt.core.ext.UnableToCompleteException;
+
+public class FallThroughGenerator extends Generator {
+    @Override
+    public String generate(TreeLogger logger, GeneratorContext context, String typeName)
+            throws UnableToCompleteException {
+        return typeName + "Impl";
+    }
+}

--- a/gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/rebind/resource/AbstractResourceGenerator.java
+++ b/gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/rebind/resource/AbstractResourceGenerator.java
@@ -45,7 +45,8 @@ public abstract class AbstractResourceGenerator extends AbstractVelocityGenerato
     protected AbstractResourceGenerator(
             Logger logger,
             GeneratorContext context,
-            VelocityEngine velocityEngine, Set<MethodGenerator> methodGenerators) {
+            VelocityEngine velocityEngine,
+            Set<MethodGenerator> methodGenerators) {
         super(logger, context, velocityEngine);
 
         this.methodGenerators = methodGenerators;

--- a/gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/rebind/resource/TopLevelResourceGenerator.java
+++ b/gwtp-core/gwtp-dispatch-rest/src/main/java/com/gwtplatform/dispatch/rest/rebind/resource/TopLevelResourceGenerator.java
@@ -1,4 +1,4 @@
-/**
+/*
  * Copyright 2014 ArcBees Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not

--- a/gwtp-core/gwtp-dispatch-rest/src/main/resources/com/gwtplatform/dispatch/rest/DispatchRest.gwt.xml
+++ b/gwtp-core/gwtp-dispatch-rest/src/main/resources/com/gwtplatform/dispatch/rest/DispatchRest.gwt.xml
@@ -27,4 +27,12 @@
     <generate-with class="com.gwtplatform.dispatch.rest.rebind.DispatchRestIncrementalGenerator">
         <when-type-is class="com.gwtplatform.dispatch.rest.client.DispatchRestEntryPoint"/>
     </generate-with>
+
+    <!-- Fallbacks -->
+    <generate-with class="com.gwtplatform.dispatch.rest.rebind.FallThroughGenerator">
+        <when-type-assignable class="com.gwtplatform.dispatch.rest.client.annotations.Resource"/>
+    </generate-with>
+    <generate-with class="com.gwtplatform.dispatch.rest.rebind.FallThroughGenerator">
+        <when-type-is class="com.gwtplatform.dispatch.rest.client.serialization.JacksonMapperProvider"/>
+    </generate-with>
 </module>

--- a/gwtp-core/gwtp-dispatch-rest/src/main/resources/com/gwtplatform/dispatch/rest/rebind/resource/Resource.vm
+++ b/gwtp-core/gwtp-dispatch-rest/src/main/resources/com/gwtplatform/dispatch/rest/rebind/resource/Resource.vm
@@ -1,8 +1,6 @@
 package $package;
 
-import javax.inject.Inject;
-
-import com.gwtplatform.dispatch.rest.client.annotations.DefaultDateFormat;
+import com.gwtplatform.dispatch.rest.client.core.StaticParametersFactory;
 import com.gwtplatform.dispatch.rest.client.core.parameters.HttpParameterFactory;
 #foreach ($import in $imports)
 import $import;
@@ -12,12 +10,9 @@ public class $impl implements $resourceType {
     private final HttpParameterFactory httpParameterFactory;
     private final String defaultDateFormat;
 
-    @Inject
-    ${impl}(
-            HttpParameterFactory httpParameterFactory,
-            @DefaultDateFormat String defaultDateFormat) {
-        this.defaultDateFormat = defaultDateFormat;
-        this.httpParameterFactory = httpParameterFactory;
+    ${impl}() {
+        this.defaultDateFormat = StaticParametersFactory.getDefaultDateFormat();
+        this.httpParameterFactory = StaticParametersFactory.getHttpParameterFactory();
     }
 
 #foreach ($method in $methods)

--- a/gwtp-core/gwtp-dispatch-rest/src/main/resources/com/gwtplatform/dispatch/rest/rebind/serialization/JacksonMapperProvider.vm
+++ b/gwtp-core/gwtp-dispatch-rest/src/main/resources/com/gwtplatform/dispatch/rest/rebind/serialization/JacksonMapperProvider.vm
@@ -3,16 +3,29 @@ package $package;
 import javax.inject.Inject;
 import javax.inject.Provider;
 
+import com.github.nmorel.gwtjackson.client.ObjectMapper;
 import com.google.inject.TypeLiteral;
+import com.google.gwt.core.client.GWT;
 #if (!$definitions.isEmpty())
 import ${package}.mappers.*;
 #end
 
 public class $impl extends AbstractJacksonMapperProvider {
-    @Inject
-    $impl(#foreach ($definition in $definitions)Provider<$definition.getParameterizedClassName()> ${definition.className}Provider#commaIfNeeded($definitions)$lf#end) {
+    $impl() {
 #foreach ($definition in $definitions)
-        registerProvider("$definition.serializableType.getParameterizedQualifiedSourceName()", ${definition.className}Provider);
+        registerProvider("$definition.serializableType.getParameterizedQualifiedSourceName()",
+                new Provider<ObjectMapper<?>>() {
+                    private ObjectMapper<?> mapper;
+
+                    @Override
+                    public ObjectMapper<?> get() {
+                        if (mapper == null){
+                            mapper = (ObjectMapper<?>) GWT.create(${definition.className}.class);
+                        }
+
+                        return mapper;
+                    }
+                });
 #end
     }
 }

